cardano-testnet/test/cardano-testnet-test/Cardano/Testnet/Test/Cli/Query.hs

```diff
@@ -78,101 +112,280 @@ hprop_cli_queries = integrationWorkspace "cli-queries" $ \tempAbsBasePath' -> H.
+      -- constitution
+      do
+        output <- execCli' execConfig [ eraName, "query", "constitution" ]
+        H.diffVsGoldenFile output "test/cardano-testnet-test/files/golden/queries/queryConstitutionOut.json"
+
+    TestQueryGovStateCmd ->
+      -- gov-state
+      do
+        -- wait for the proposal stage to end
+        shelleyGenesisVal <- H.evalEitherM $ readJsonFile shelleyGeneisFile
+        newSlot <- waitForFuturePParamsToStabilise epochStateView shelleyGenesisVal
+        H.note_ $ "Current slot is: " ++ show newSlot
+        -- to stdout
+        output <- execCli' execConfig [ eraName, "query", "gov-state" ]
+        H.diffVsGoldenFile output "test/cardano-testnet-test/files/golden/queries/govStateOut.json"
+        -- to a file
+        let govStateOutFile = work </> "gov-state-out.json"
+        H.noteM_ $ execCli' execConfig [ eraName, "query", "gov-state", "--out-file", govStateOutFile ]
+        H.diffFileVsGoldenFile
+          govStateOutFile
+          "test/cardano-testnet-test/files/golden/queries/govStateOut.json"
+
+    TestQueryDRepStateCmd ->
+      -- drep-state
+      do
+        -- to stdout
+        -- TODO: deserialize to a Haskell value when
+        -- https://github.com/IntersectMBO/cardano-cli/issues/606 is tackled
+        dreps :: Aeson.Value <- H.noteShowM $ execCliStdoutToJson execConfig [ eraName, "query", "drep-state", "--all-dreps" ]
+        assertArrayOfSize dreps 3
+        -- to a file
+        let drepStateOutFile = work </> "drep-state-out.json"
+        H.noteM_ $ execCli' execConfig [ eraName, "query", "drep-state", "--all-dreps"
+                                       , "--out-file", drepStateOutFile ]
+        _ :: Aeson.Value <- H.readJsonFileOk drepStateOutFile
+        pure ()
+
+    TestQueryDRepStakeDistributionCmd ->
+      -- drep-stake-distribution
+      H.noteM_ $ execCli' execConfig [ eraName, "query", "drep-stake-distribution", "--all-dreps" ]
+
+    TestQueryCommitteeMembersStateCmd ->
+      -- committee-state
+      H.noteM_ $ execCli' execConfig [ eraName, "query", "committee-state" ]
+
+    TestQueryTreasuryValueCmd -> do
+      -- treasury
+      H.noteM_ $ execCli' execConfig [ eraName, "query", "treasury" ]
+
+  where
+  -- | Wait for the part of the epoch when futurePParams are known
+  waitForFuturePParamsToStabilise
+    :: HasCallStack
+    => MonadIO m
+    => MonadTest m
+    => MonadAssertion m
+    => MonadCatch m
+    => EpochStateView
+    -> ShelleyGenesis StandardCrypto
+    -> m SlotNo -- ^ The block number reached
+  waitForFuturePParamsToStabilise epochStateView shelleyGenesisConf = withFrozenCallStack $
+    H.noteShowM . H.nothingFailM $
+      watchEpochStateUpdate epochStateView (EpochInterval 2) $ \(_, slotNo, _) -> do
+        pure $ if areFuturePParamsStable shelleyGenesisConf slotNo
+               then Just slotNo
+               else Nothing
+
+  areFuturePParamsStable :: ShelleyGenesis StandardCrypto -> SlotNo -> Bool
```

Comment:
Why do we need to do this check? Won't the params be the same as the ones you've set in testnet setup in line 84?

Reply:
Yes, this has to do with the non-determinism of the `futurePParams` field in the result of `gov-state` query. Apparently it is not possible to know for sure what the `futurePParams` (for the next epoch) are going to be until they get settled, and that happens after the slot `4 * securityParam / slotCoeff` of each epoch. If we query before that we get `PotentialPParamsUpdate`, and that messes with the golden file.

You can see [this discussion in slack](https://input-output-rnd.slack.com/archives/CCRB7BU8Y/p1722515058261859) for more info

Reply:
I added a comment here: https://github.com/IntersectMBO/cardano-node/commit/63c852a5823b666d372e585b2c5bb4921bd6b016